src/include/function/gds/gds.h

```diff
@@ -1,129 +1,80 @@
 
 namespace function {
 
-class PathLengths;
-
-struct GDSBindInput {
-    binder::expression_vector params;
-    binder::expression_vector optionalParams;
-    binder::Binder* binder = nullptr;
-    std::vector<parser::YieldVariable> yieldVariables;
-
-    GDSBindInput() = default;
-
-    std::shared_ptr<binder::Expression> getParam(common::idx_t idx) const { return params[idx]; }
-    std::shared_ptr<binder::Expression> getOptionalParam(common::idx_t idx) const {
-        return optionalParams[idx];
-    }
-    common::idx_t getNumParams() const { return params.size(); }
-};
-
-struct GDSBindData {
+struct GDSBindData : public TableFuncBindData {
     graph::GraphEntry graphEntry;
     std::shared_ptr<binder::Expression> nodeOutput;
 
-    GDSBindData(graph::GraphEntry graphEntry, std::shared_ptr<binder::Expression> nodeOutput)
-        : graphEntry{graphEntry.copy()}, nodeOutput{std::move(nodeOutput)} {}
+    GDSBindData(binder::expression_vector columns, graph::GraphEntry graphEntry,
+        std::shared_ptr<binder::Expression> nodeOutput)
+        : TableFuncBindData{std::move(columns)}, graphEntry{graphEntry.copy()},
+          nodeOutput{std::move(nodeOutput)} {}
     GDSBindData(const GDSBindData& other)
-        : graphEntry{other.graphEntry.copy()}, nodeOutput{other.nodeOutput} {}
-
-    virtual ~GDSBindData() = default;
+        : TableFuncBindData{other}, graphEntry{other.graphEntry.copy()},
+          nodeOutput{other.nodeOutput}, resultTable{other.resultTable} {}
 
-    virtual bool hasNodeInput() const { return false; }
-
-    virtual std::shared_ptr<binder::Expression> getNodeInput() const { return nullptr; }
-
-    virtual bool hasNodeOutput() const { return false; }
-
-    virtual std::shared_ptr<binder::Expression> getNodeOutput() const {
-        KU_ASSERT(nodeOutput != nullptr);
-        return nodeOutput;
+    void setResultFTable(std::shared_ptr<processor::FactorizedTable> table) {
+        resultTable = std::move(table);
     }
+    std::shared_ptr<processor::FactorizedTable> getResultTable() const { return resultTable; }
 
-    virtual std::unique_ptr<GDSBindData> copy() const {
+    std::unique_ptr<TableFuncBindData> copy() const override {
         return std::make_unique<GDSBindData>(*this);
     }
 
-    template<class TARGET>
-    TARGET* ptrCast() {
-        return common::ku_dynamic_cast<TARGET*>(this);
-    }
+private:
+    std::shared_ptr<processor::FactorizedTable> resultTable;
 };
 
-// Base class for every graph data science algorithm.
-class KUZU_API GDSAlgorithm {
-protected:
-    static constexpr char NODE_COLUMN_NAME[] = "node";
+struct KUZU_API GDSFuncSharedState : public TableFuncSharedState {
+    std::unique_ptr<graph::Graph> graph;
 
-public:
-    GDSAlgorithm() = default;
+    GDSFuncSharedState(std::shared_ptr<processor::FactorizedTable> fTable,
+        std::unique_ptr<graph::Graph> graph)
+        : TableFuncSharedState{}, graph{std::move(graph)}, factorizedTablePool{std::move(fTable)} {}
```

Comment:
it is better to pass the `numRows` to TableFuncSharedState because it is used by hashindex to reserve slots during copying.

Reply:
ditto. If the table function is not meant to be used as a scan source, I don't think numRows actually matters.